File: backend-node/src/services/recurring.ts
/** Recurring rules and the pass that turns them into transactions. */

import { randomUUID } from 'node:crypto'
import { notFound, validationError } from '../core/errors.js'
import { addMonths, addYears } from '../domain/dates.js'
import { Frequency, type EnumValue, TransactionType } from '../domain/enums.js'
import { Instant } from '../domain/instant.js'
import type { Money } from '../domain/money.js'
import {
  AccountRepository,
  type RecurringRule,
  RecurringRepository,
  type Transaction,
  TransactionRepository,
} from '../repositories/index.js'
import type { CategoryService } from './index.js'

const RECURRING_ENTITY = 'Recurring rule'
const RECURRING_TRANSFER_MESSAGE = 'Recurring transfers are not supported.'
const RECURRING_END_DATE_MESSAGE = 'End date must not be before the start date.'
const ACCOUNT_ENTITY = 'Account'

/**
 * A cap so one long-dormant rule cannot generate an unbounded batch in a single
 * pass; the rule stays active and the next pass carries on.
 */
export const MAX_OCCURRENCES_PER_PASS = 500
const RECURRING_TAG = 'recurring'

export function recurringDto(rule: RecurringRule): Record<string, unknown> {
  return {
    id: rule.id,
    accountId: rule.accountId,
    categoryId: rule.categoryId,
    type: rule.type,
    amount: rule.amount,
    description: rule.description,
    frequency: rule.frequency,
    startDate: rule.startDate,
    endDate: rule.endDate,
    nextRunDate: rule.nextRunDate,
    isActive: rule.isActive,
  }
}

/** The next occurrence after `moment`. Clock time is preserved throughout. */
export function advance(moment: Date, frequency: EnumValue): Date {
  if (frequency.equals(Frequency.Daily)) {
    return new Date(moment.getTime() + 24 * 60 * 60 * 1000)
  }
  if (frequency.equals(Frequency.Weekly)) {
    return new Date(moment.getTime() + 7 * 24 * 60 * 60 * 1000)
  }
  if (frequency.equals(Frequency.Yearly)) return addYears(moment, 1)
  // Monthly, and anything undefined, advance by a clamped month.
  return addMonths(moment, 1)
}

/** Emits the occurrences due at or before the cutoff, mutating the rule. */
export function materialize(rule: RecurringRule, cutoff: Instant): Transaction[] {
  const created: Transaction[] = []

  for (let index = 0; index < MAX_OCCURRENCES_PER_PASS; index += 1) {
    if (!rule.isActive || rule.nextRunDate.valueOf() > cutoff.valueOf()) break
    if (rule.endDate !== null && rule.nextRunDate.valueOf() > rule.endDate.valueOf()) {
      // An occurrence exactly *on* the end date is still created; only a run
      // past it retires the rule.
      rule.isActive = false
      break
    }

    created.push({
      id: randomUUID(),
      userId: rule.userId,
      accountId: rule.accountId,
      categoryId: rule.categoryId,
      type: rule.type,
      amount: rule.amount,
      date: rule.nextRunDate,
      description: rule.description,
      notes: null,
      tags: [RECURRING_TAG],
      transferAccountId: null,
    })

    rule.nextRunDate = Instant.fromDate(advance(rule.nextRunDate.date, rule.frequency))
  }

  return created
}

export interface RecurringInput {
  accountId: string
  categoryId: string | null
  type: EnumValue
  amount: Money
  description: string
  frequency: EnumValue
  startDate: Instant
  endDate: Instant | null
  isActive: boolean | null
}

export class RecurringService {
  constructor(
    private readonly rules: RecurringRepository,
    private readonly transactions: TransactionRepository,
    private readonly accounts: AccountRepository,
    private readonly categories: CategoryService,
  ) {}

  async listAll(userId: string) {
    return (await this.rules.listAll(userId)).map(recurringDto)
  }

  async create(userId: string, input: RecurringInput) {
    await this.check(userId, input)

    const rule: RecurringRule = {
      id: randomUUID(),
      userId,
      ...shape(input),
      // The first occurrence is the start date itself.
      nextRunDate: input.startDate,
    }
    await this.rules.add(rule)
    return recurringDto(rule)
  }

  async update(userId: string, ruleId: string, input: RecurringInput) {
    await this.check(userId, input)
    const rule = await this.load(userId, ruleId)
    Object.assign(rule, shape(input))

    // Pull the next run forward if the start moved later; never push it back.
    if (rule.nextRunDate.valueOf() < rule.startDate.valueOf()) {
      rule.nextRunDate = rule.startDate
    }

    await this.rules.update(rule)
    return recurringDto(rule)
  }

  async remove(userId: string, ruleId: string): Promise<void> {
    if (!(await this.rules.remove(userId, ruleId))) throw notFound(RECURRING_ENTITY)
  }

  /** Runs one pass. The caller owns the transaction and the lock. */
  async materializeDue(now: Instant): Promise<number> {
    const due = await this.rules.listDue(now)

    const created: Transaction[] = []
    for (const rule of due) {
      created.push(...materialize(rule, now))
    }

    if (created.length > 0) await this.transactions.addMany(created)
    for (const rule of due) {
      await this.rules.update(rule)
    }

    return created.length
  }

  /** Order fixed: it decides which error a doubly-wrong request gets. */
  private async check(userId: string, input: RecurringInput): Promise<void> {
    if (input.type.equals(TransactionType.Transfer)) {
      throw validationError(RECURRING_TRANSFER_MESSAGE)
    }
    if (!(await this.accounts.exists(userId, input.accountId))) throw notFound(ACCOUNT_ENTITY)
    if (input.endDate !== null && input.endDate.valueOf() < input.startDate.valueOf()) {
      throw validationError(RECURRING_END_DATE_MESSAGE)
    }
    await this.categories.ensureUsable(userId, input.categoryId)
  }

  private async load(userId: string, ruleId: string): Promise<RecurringRule> {
    const rule = await this.rules.get(userId, ruleId)
    if (!rule) throw notFound(RECURRING_ENTITY)
    return rule
  }
}

function shape(input: RecurringInput) {
  return {
    accountId: input.accountId,
    categoryId: input.categoryId,
    type: input.type,
    amount: input.amount.round(),
    description: input.description.trim(),
    frequency: input.frequency,
    startDate: input.startDate,
    endDate: input.endDate,
    // An omitted flag leaves the rule running.
    isActive: input.isActive ?? true,
  }
}

File: backend-py/app/repositories/recurring.py
"""Recurring rules. ListDue crosses users: the worker runs one pass for everyone."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal

import asyncpg

from app.domain.enums import Frequency, TransactionType

_COLUMNS = (
    '"Id", "UserId", "AccountId", "CategoryId", "Type", "Amount", "Description", '
    '"Frequency", "StartDate", "EndDate", "NextRunDate", "IsActive"'
)


@dataclass(slots=True)
class RecurringRule:
    id: uuid.UUID
    user_id: uuid.UUID
    account_id: uuid.UUID
    category_id: uuid.UUID | None
    type: TransactionType
    amount: Decimal
    description: str
    frequency: Frequency
    start_date: datetime
    end_date: datetime | None
    next_run_date: datetime
    is_active: bool


def _row(record: asyncpg.Record) -> RecurringRule:
    return RecurringRule(
        id=record["Id"],
        user_id=record["UserId"],
        account_id=record["AccountId"],
        category_id=record["CategoryId"],
        type=TransactionType(record["Type"]),
        amount=record["Amount"],
        description=record["Description"],
        frequency=Frequency(record["Frequency"]),
        start_date=record["StartDate"],
        end_date=record["EndDate"],
        next_run_date=record["NextRunDate"],
        is_active=record["IsActive"],
    )


class RecurringRepository:
    def __init__(self, conn: asyncpg.Connection) -> None:
        self._conn = conn

    async def list_all(self, user_id: uuid.UUID) -> list[RecurringRule]:
        records = await self._conn.fetch(
            f'SELECT {_COLUMNS} FROM "RecurringRules" WHERE "UserId" = $1 ORDER BY "NextRunDate"',
            user_id,
        )
        return [_row(record) for record in records]

    async def get(self, user_id: uuid.UUID, rule_id: uuid.UUID) -> RecurringRule | None:
        record = await self._conn.fetchrow(
            f'SELECT {_COLUMNS} FROM "RecurringRules" WHERE "Id" = $1 AND "UserId" = $2',
            rule_id,
            user_id,
        )
        return _row(record) if record else None

    async def list_due(self, cutoff: datetime) -> list[RecurringRule]:
        """Every active rule due at or before the cutoff, across all users."""
        records = await self._conn.fetch(
            f'SELECT {_COLUMNS} FROM "RecurringRules" '
            'WHERE "IsActive" = true AND "NextRunDate" <= $1 ORDER BY "NextRunDate", "Id"',
            cutoff,
        )
        return [_row(record) for record in records]

    async def add(self, rule: RecurringRule) -> None:
        await self._conn.execute(
            'INSERT INTO "RecurringRules" ("Id", "UserId", "AccountId", "CategoryId", '
            '"Type", "Amount", "Description", "Frequency", "StartDate", "EndDate", '
            '"NextRunDate", "IsActive") '
            "VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)",
            rule.id,
            rule.user_id,
            rule.account_id,
            rule.category_id,
            int(rule.type),
            rule.amount,
            rule.description,
            int(rule.frequency),
            rule.start_date,
            rule.end_date,
            rule.next_run_date,
            rule.is_active,
        )

    async def update(self, rule: RecurringRule) -> None:
        await self._conn.execute(
            'UPDATE "RecurringRules" SET "AccountId" = $3, "CategoryId" = $4, "Type" = $5, '
            '"Amount" = $6, "Description" = $7, "Frequency" = $8, "StartDate" = $9, '
            '"EndDate" = $10, "NextRunDate" = $11, "IsActive" = $12 '
            'WHERE "Id" = $1 AND "UserId" = $2',
            rule.id,
            rule.user_id,
            rule.account_id,
            rule.category_id,
            int(rule.type),
            rule.amount,
            rule.description,
            int(rule.frequency),
            rule.start_date,
            rule.end_date,
            rule.next_run_date,
            rule.is_active,
        )

    async def delete(self, user_id: uuid.UUID, rule_id: uuid.UUID) -> bool:
        status = await self._conn.execute(
            'DELETE FROM "RecurringRules" WHERE "Id" = $1 AND "UserId" = $2', rule_id, user_id
        )
        return status != "DELETE 0"
